File: crates/book-forge-converter/src/metadata.rs
use serde::{Deserialize, Serialize};

use crate::{BookMetadata, text::collapse_whitespace, text::strip_html_tags};

const MODIFIED_DATE: &str = "2026-05-10T00:00:00Z";

#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
pub struct SanitizedMetadata {
    pub title: String,
    pub author: String,
    pub language: String,
    pub description: String,
    pub identifier: String,
    pub modified: String,
}

pub(crate) fn sanitize_metadata(metadata: BookMetadata, source_url: &str) -> SanitizedMetadata {
    let title = sanitize_metadata_value(&metadata.title, "Untitled Book");
    let author = sanitize_metadata_value(&metadata.author, "Unknown Author");
    let language = sanitize_language(&metadata.language);
    let description = sanitize_metadata_value(&metadata.description, "");
    let identifier = stable_identifier(source_url, &title);

    SanitizedMetadata {
        title,
        author,
        language,
        description,
        identifier,
        modified: MODIFIED_DATE.to_string(),
    }
}

pub(crate) fn sanitize_metadata_value(raw: &str, fallback: &str) -> String {
    let without_tags = strip_html_tags(raw);
    let mut cleaned = String::with_capacity(without_tags.len());

    for character in without_tags.chars() {
        if character.is_control()
            || matches!(
                character,
                '/' | '\\' | '\u{2028}' | '\u{2029}' | '<' | '>' | '"' | '\''
            )
        {
            cleaned.push(' ');
        } else {
            cleaned.push(character);
        }
    }

    while cleaned.contains("..") {
        cleaned = cleaned.replace("..", " ");
    }

    let collapsed = collapse_whitespace(&cleaned);
    if collapsed.is_empty() {
        fallback.to_string()
    } else {
        collapsed
    }
}

pub(crate) fn safe_download_filename(title: &str) -> String {
    let mut filename = String::with_capacity(title.len() + 5);
    let mut previous_separator = false;

    for character in title.chars() {
        if character.is_ascii_alphanumeric() {
            filename.push(character.to_ascii_lowercase());
            previous_separator = false;
        } else if !previous_separator {
            filename.push('-');
            previous_separator = true;
        }

        if filename.len() >= 80 {
            break;
        }
    }

    let filename = filename.trim_matches('-').trim_matches('.').to_string();
    let filename = if filename.is_empty() {
        "book-forge".to_string()
    } else {
        filename
    };

    format!("{filename}.epub")
}

fn sanitize_language(raw: &str) -> String {
    let candidate: String = raw
        .trim()
        .chars()
        .filter(|character| character.is_ascii_alphanumeric() || *character == '-')
        .take(35)
        .collect();

    if candidate.is_empty() {
        "en".to_string()
    } else {
        candidate
    }
}

fn stable_identifier(source_url: &str, title: &str) -> String {
    let mut hash = 0xcbf29ce484222325_u64;
    for byte in source_url.bytes().chain(title.bytes()) {
        hash ^= u64::from(byte);
        hash = hash.wrapping_mul(0x100000001b3);
    }
    format!("urn:book-forge:{hash:016x}")
}

File: converter/metadata.go
package converter

import (
	"fmt"
	"strings"
)

const modifiedDate = "2026-05-10T00:00:00Z"

type SanitizedMetadata struct {
	Title       string `json:"title"`
	Author      string `json:"author"`
	Language    string `json:"language"`
	Description string `json:"description"`
	Identifier  string `json:"identifier"`
	Modified    string `json:"modified"`
}

func sanitizeMetadata(metadata BookMetadata, sourceURL string) SanitizedMetadata {
	title := sanitizeMetadataValue(metadata.Title, "Untitled Book")
	author := sanitizeMetadataValue(metadata.Author, "Unknown Author")
	language := sanitizeLanguage(metadata.Language)
	description := sanitizeMetadataValue(metadata.Description, "")
	identifier := stableIdentifier(sourceURL, title)

	return SanitizedMetadata{
		Title:       title,
		Author:      author,
		Language:    language,
		Description: description,
		Identifier:  identifier,
		Modified:    modifiedDate,
	}
}

func sanitizeMetadataValue(raw, fallback string) string {
	withoutTags := stripHTMLTags(raw)
	var cleaned strings.Builder
	cleaned.Grow(len(withoutTags))

	for _, ch := range withoutTags {
		if isControl(ch) || strings.ContainsRune("/\\<>\"'", ch) || ch == '\u2028' || ch == '\u2029' {
			cleaned.WriteByte(' ')
		} else {
			cleaned.WriteRune(ch)
		}
	}

	result := cleaned.String()
	for strings.Contains(result, "..") {
		result = strings.ReplaceAll(result, "..", " ")
	}

	collapsed := collapseWhitespace(result)
	if collapsed == "" {
		return fallback
	}
	return collapsed
}

func SafeDownloadFilename(title string) string {
	var filename strings.Builder
	filename.Grow(len(title) + 5)
	previousSep := false

	for _, ch := range title {
		if len([]byte(filename.String())) >= 80 {
			break
		}
		if ch >= 'a' && ch <= 'z' || ch >= 'A' && ch <= 'Z' || ch >= '0' && ch <= '9' {
			filename.WriteByte(byte(ch | 32))
			previousSep = false
		} else if !previousSep {
			filename.WriteByte('-')
			previousSep = true
		}
	}

	name := strings.Trim(filename.String(), "-.")
	if name == "" {
		name = "book-forge"
	}
	return name + ".epub"
}

func sanitizeLanguage(raw string) string {
	var candidate strings.Builder
	for _, ch := range raw {
		if candidate.Len() >= 35 {
			break
		}
		if (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') || ch == '-' {
			candidate.WriteRune(ch)
		}
	}
	result := strings.TrimSpace(candidate.String())
	if result == "" {
		return "en"
	}
	return result
}

func stableIdentifier(sourceURL, title string) string {
	var hash uint64 = 0xcbf29ce484222325
	for _, b := range []byte(sourceURL) {
		hash ^= uint64(b)
		hash *= 0x100000001b3
	}
	for _, b := range []byte(title) {
		hash ^= uint64(b)
		hash *= 0x100000001b3
	}
	return fmt.Sprintf("urn:book-forge:%016x", hash)
}

func isControl(ch rune) bool {
	return ch < 32 && ch != '\n' && ch != '\r' && ch != '\t'
}
